### PR TITLE
fix: crane image pull fail due to rate limit & migration test 

### LINKED ISSUE #4783: hardhat test in main branch is failing
### Describe the bug

https://github.com/hiero-ledger/solo/actions/runs/28009718147/job/82978454032

### Describe the expected behavior

Should not fail 

### To Reproduce

PR checks

### Additional Context

_No response_

### LINKED ISSUE #4801: One-Shot Timing workflow relay startup can fail under low memory limit
## Summary
The One-Shot Timing workflow can fail during the cold start path while deploying the JSON-RPC relay. In run https://github.com/hiero-ledger/solo/actions/runs/28038802362/job/82999317871, `solo relay node add` installed `relay-1`, but the relay pod never became ready and the workflow failed with `Relay relay-1 is not ready`.

## Evidence
- Workflow: One-Shot Timing Test
- Job: `one-shot-timing`
- Run/job URL: https://github.com/hiero-ledger/solo/actions/runs/28038802362/job/82999317871
- Failure surfaced as `Relay relay-1 is not ready: No pod found for: pods:app.kubernetes.io/instance=relay-1,app.kubernetes.io/name=relay`.
- Uploaded diagnostics showed the relay pod existed but failed startup readiness probes on `/health/readiness`.
- Relay container status showed restart count 2 and a prior terminated state with exit code 137.
- The deployed HTTP relay resources were capped lower than the websocket relay resources.

## Proposed Fix
Increase the default HTTP relay memory limit in `resources/relay-values.yaml` so the relay is not killed during startup in the default one-shot timing path.

## Acceptance Criteria
- The one-shot timing workflow can complete the relay startup phase without the relay container being killed for memory pressure.
- The default one-shot deployment keeps parallel deploy behavior intact.

### PR DESCRIPTION
**Pull request by JeffreyDallas**
## Description

This pull request changes the following:

* Updates the default Kind Docker Hub mirror order so `hub.mirror.docker.lat.ope.eng.hashgraph.io` is tried before `mirror.gcr.io` and direct Docker Hub.
* Updates Docker Hub image cache archive saving to try Kind-configured Docker Hub mirrors before direct Docker Hub registries.
* Adds an optional `KIND_DOCKER_REGISTRY_MIRRORS` override for explicit cache-pull source configuration and unit tests.
* Throws cache archive save failures only when the error or its cause chain indicates Docker Hub rate limiting; non-rate-limit save failures continue without registering that cache item.
* Increases the Hardhat example timeout for Solo-backed tests to a configurable 300 seconds and caps the Hardhat test process at 10 minutes so failures do not hang the workflow step.
* Raises the relay memory limit to `700Mi` for Solo CI relay stability.
* Fixes migration workflow source deployment for block-node scenarios by generating source `application.properties`, one-shot values, and mirror-node values consumed by the released Solo initial deploy:
  * consensus node versions before `v0.74.0` use `blockStream.streamMode=BOTH`, MinIO enabled through `setup --minio`, and mirror record-stream ingestion.
  * consensus node versions `v0.74.0` and newer use `blockStream.streamMode=BLOCKS`, MinIO disabled through `setup --minio false`, and mirror block-node ingestion.
* Raises the external ERC20 migration smoke-test timeout to 300 seconds by patching the cloned `hedera-smart-contracts` test locally, with `CONTRACT_TEST_TIMEOUT_MS` override support.
* Adds unit coverage for image-cache mirror resolution and rate-limit failure handling.

Root cause: the original image-cache failure was Docker Hub rate limiting in the image cache path. `crane pull` does not consume Kind/containerd registry mirror configuration, so direct Docker Hub pulls hit anonymous rate limits for images such as `docker.io/library/busybox` and `docker.io/library/ubuntu`.

The later migration failures came from the initial deployment using released Solo, so local source changes do not affect its generated one-shot argv. The migration workflow now passes the needed stream mode, MinIO choice, and mirror importer mode through generated values files consumed by the released Solo one-shot deploy. Pre-`v0.74.0` source networks still need record stream ingestion backed by MinIO, while `v0.74.0` and newer block-node source networks can disable MinIO and use mirror block-node ingestion.

Shared GitHub Actions cache warm-up work was split out to branch `ci-share-solo-image-cache` and is tracked separately in #4803.

### Related Issues

* Closes #4783
* Closes #4801
* Related to #4803

### Pull request (PR) checklist

* [x] This PR added tests (unit, integration, and/or end-to-end)
* [ ] This PR updated documentation
* [x] This PR added no TODOs or commented out code
* [x] This PR has no breaking changes
* [x] Any technical debt has been documented as a separate issue and linked to this PR
* [x] Any `package.json` changes have been explained to and approved by a repository manager
* [x] All related issues have been linked to this PR
* [x] All changes in this PR are included in the description
* [x] When this PR merges the commits will be squashed and the title will be used as the commit message, the 'commit message guidelines' below have been followed

### Testing

* [x] This PR added unit tests
* [ ] This PR added integration/end-to-end tests
* [x] These changes required manual testing that is documented below
* [x] Anything not tested is documented

The following manual testing was done:

* `npx mocha 'test/unit/integration/cache/image-cache-handler.test.ts'`
* `npx mocha 'test/unit/commands/one-shot/orchestrator/deploy/deploy-argv-builders.test.ts'` (`36 passing`, after reverting the ineffective local argv-builder changes)
* `bash -n .github/workflows/script/launch_network.sh`
* `bash -n .github/workflows/script/solo_smoke_test.sh`
* `task build:compile`
* `task format` (completed with the existing warning backlog)
* `npx eslint src/integration/cache/impl/image-cache-handler.ts test/unit/integration/cache/image-cache-handler.test.ts`
* `npx hardhat compile` from `examples/hardhat-with-solo/hardhat-example`
* `SOLO_HARDHAT_TEST_TIMEOUT_MS=300000 timeout 60s npx hardhat test mocha --bail --network my_solo_deployment --grep no-tests-should-match` from `examples/hardhat-with-solo/hardhat-example` to validate Hardhat 3 command parsing without a live Solo relay
* `git diff --check`
* `for cluster in $(kind get clusters); do kind delete cluster -n "$cluster"; done; rm -rf ~/.solo/*; .github/workflows/script/launch_network.sh "0.79.0" "v0.74.0"` passed locally before the cleanup that removed redundant manual `cluster-ref config` calls.

The following was not tested:

* The full migration workflow has not been rerun locally after removing the redundant manual `cluster-ref config connect/setup` calls; one-shot should perform those steps from the generated values file.
* The GitHub Actions migration workflow has not completed after the latest ERC20 timeout patch.
* The full Hardhat example was not rerun locally against a live Solo deployment after increasing the timeout and adding the 10-minute process cap.
* No live registry/network integration test was run specifically against the Hashgraph Docker Hub mirror.
* Shared GitHub Actions cache warm-up is not included in this PR; it is tracked by #4803 and branch `ci-share-solo-image-cache`.
* `npx tsc --noEmit` in the Hardhat example was attempted earlier, but the example has existing Hardhat 3 module-resolution/type errors unrelated to this change.

<details>
<summary>
Commit message guidelines
</summary>
We use 'Conventional Commits' to ensure that our commit messages are easy to read, follow a consistent format, and for automated release note generation. Please follow the guidelines below when writing your commit messages:

1. BREAKING CHANGE: a commit that has a footer BREAKING CHANGE:, or appends a ! after the type/scope, introduces a breaking API change (correlating with MAJOR in Semantic Versioning). A BREAKING CHANGE can be part of commits of any type.  NOTE: currently breaking changes will only bump the MAJOR version.
2. The title is prefixed with one of the following:

| Prefix    | Description                                         | Semantic Version Update | Captured in Release Notes |
|-----------|-----------------------------------------------------|-------------------------|---------------------------|
| feat:     | a new feature                                       | MINOR                   | Yes                       |
| fix:      | a bug fix                                           | PATCH                   | Yes                       |
| perf:     | performance                                         | PATCH                   | Yes                       |
| refactor: | code change that isn't feature or fix               | none                    | No                        |
| test:     | adding missing tests                                | none                    | No                        |
| docs:     | changes to documentation                            | none                    | Yes                       |
| build:    | changes to build process                            | none                    | No                        |
| ci:       | changes to CI configuration                         | none                    | No                        |
| style:    | formatting, missing semi-colons, etc                | none                    | No                        |
| chore:    | updating grunt tasks etc; no production code change | none                    | No                        |

</details>

### CODE REVIEW

File: resources/relay-values.yaml [hunk 1]
```diff
@@ -27,7 +27,7 @@ relay:
       memory: 0
     limits:
       cpu: 500m
-      memory: 250Mi
+      memory: 700Mi
```
Comment (jeromy-cannon) [inline resources/relay-values.yaml:30]: this requires the new process to be followed.
https://swirldslabs.slack.com/archives/C07UC0V0Y75/p1779389699681429

Image

**Review by jeromy-cannon — COMMENTED** [review]
## Intent

Fixes Docker Hub rate-limit failures in the image-cache `crane pull` path by routing saves through Kind-configured mirrors, and patches the migration CI workflow to correctly configure block-stream mode and MinIO for both pre- and post-TSS consensus node versions.

## Verdict

Request changes — 2 majors to address before merge; minors are fine to fix in follow-up.

## Critical

None.

## Major

- **Empty catch block** — `src/integration/cache/impl/image-cache-handler.ts:289` — §4.9 bans empty catch blocks without a comment. All errors from reading/parsing the kind config (file-not-found, YAML parse failures, regex errors) are silently swallowed here, making debugging unexpectedly hard in environments where the kind config is absent or malformed. See inline for a one-line fix.

- **`process.env` instead of `getEnvironmentVariable()`** — `src/integration/cache/impl/image-cache-handler.ts:255` — the project convention for env-var access in `src/` is `getEnvironmentVariable()` (see `src/core/constants.ts:57` where `SOLO_KIND_CLUSTER_CONFIG_FILE` follows this exact pattern). Using `process.env` directly also bypasses the documentation requirement: CLAUDE.md tracks env vars consumed via `getEnvironmentVariable()` and requires them in `env.md`. Switch to `getEnvironmentVariable('KIND_DOCKER_REGISTRY_MIRRORS')` and add the var to `env.md`.

## Minor

- **Unit test implicitly reads `kind-config.yaml` from disk** — `test/unit/integration/cache/image-cache-handler.test.ts:139` — see inline.

- **Missing SPDX headers** — `.github/workflows/script/launch_network.sh` and `solo_smoke_test.sh` — both shell scripts are missing `# SPDX-License-Identifier: Apache-2.0` (checklist item 16). With 92 new lines in `launch_network.sh`, this PR is a good moment to add it.

- **`mocha` positional argument in Taskfile** — `examples/hardhat-with-solo/Taskfile.yml:37` — in Hardhat, positional arguments to `npx hardhat test` are treated as test-file globs. `mocha` won't match `SimpleStorage.ts`, which would result in zero tests running. If this was unintentional, the fix is:
  ```
  timeout 10m npx hardhat test --bail --network my_solo_deployment
  ```

## Questions for the author

- `install_minio_operator_for_source_deploy()` embeds Helm chart installation directly in the CI script. Is there a plan to surface this through Solo's own one-shot command? Users reproducing the migration locally won't run this step automatically.

- The `sed` patch of `test/openzeppelin/ERC-20/ERC20.js` in `solo_smoke_test.sh` modifies `hedera-smart-contracts` source directly. Is there a way to configure the timeout without patching upstream source? Or a plan to upstream a timeout config option to `hedera-smart-contracts`?

## Positive

- The mirror fallback chain (`KIND_DOCKER_REGISTRY_MIRRORS` override → kind-config parsing → `docker.io` → `registry-1.docker.io`) is well-layered and degrades gracefully.
- `isRateLimitError`'s depth-capped cause-chain traversal is a nice defensive pattern.
- The tests cover both the "rate limit → propagate" and "non-rate-limit → swallow" paths clearly, with precise assertions on call counts and argument values.
- Using `is_tss_supported_consensus_version()` to gate both block-stream mode and MinIO choice is cleaner than the earlier hard-coded override env vars.

## Companion work

- solo-docs PR: N/A (no user-facing docs changed, but `env.md` update is needed per the major finding above — that file is in-repo)
- solo-containers PR: N/A

**Comment by JeffreyDallas** [comment]
Close to split to multiple smaller PRs
